satpy/dataset.py

```diff
@@ -200,112 +372,290 @@ class DatasetID(DatasetID):
-        elif isinstance(other, numbers.Number) or \
-                isinstance(other, (tuple, list)) and len(other) == 3:
-            return self.wavelength_match(self.wavelength, other)
-        else:
-            return super(DatasetID, self).__eq__(other)
+        """Compare the DatasetIDs."""
+        sdict = self._asdict()
+        try:
+            odict = other._asdict()
+        except AttributeError:
+            return False
+        for key, val in sdict.items():
+            if key in odict and odict[key] != val and val is not None:
+                return False
+        return True
 
     def __hash__(self):
-        """Generate the hash of the ID."""
-        return tuple.__hash__(self)
+        """Hash."""
+        fields = []
+        values = []
+        for field, value in sorted(self._dict.items()):
+            if value != '*':
+                fields.append(field)
+                if isinstance(value, (list, set)):
+                    value = tuple(value)
+                values.append(value)
+        return hash(tuple(zip(fields, values)))
+
+    def get(self, key, default=None):
+        """Get an item."""
+        return self._dict.get(key, default)
 
     @classmethod
-    def from_dict(cls, d, **kwargs):
+    def from_dict(cls, the_dict):
         """Convert a dict to an ID."""
-        args = []
-        for k in DATASET_KEYS:
-            val = kwargs.get(k, d.get(k))
-            # force modifiers to tuple
-            if k == 'modifiers' and val is not None:
-                val = tuple(val)
-            args.append(val)
+        return cls(**the_dict)
 
-        return cls(*args)
+    def _asdict(self):
+        return dict(zip(self._fields, self._values))
 
     def to_dict(self, trim=True):
         """Convert the ID to a dict."""
         if trim:
             return self._to_trimmed_dict()
         else:
-            return dict(zip(DATASET_KEYS, self))
+            return self._asdict()
 
     def _to_trimmed_dict(self):
-        return {key: getattr(self, key) for key in DATASET_KEYS
-                if getattr(self, key) is not None}
-
-
-def create_filtered_dsid(dataset_key, **dfilter):
-    """Create a DatasetID matching *dataset_key* and *dfilter*.
+        return {key: val for key, val in self._dict.items()
+                if val != '*'}
+
+    def __repr__(self):
+        """Represent the query."""
+        items = ("{}={}".format(key, repr(val)) for key, val in zip(self._fields, self._values))
+        return self.__class__.__name__ + "(" + ", ".join(items) + ")"
+
+    def filter_dsids(self, dsid_container):
+        """Filter datasetids based on this query."""
+        keys = iter(dsid_container)
+        for key, val in self._dict.items():
+            if val != '*':
+                keys = [k for k in keys
+                        if k.get(key) == val]
+        return keys
+
+    def sort_dsids(self, dsids):
+        """Sort the datasetids based on this query.
```

Comment:
Sort `DataID`s (if I'm correctly understanding that's what it does).

If the term `DatasetID` is retired from Satpy then the argument `dsids` may be confusing and is perhaps better replaced by `dataids` (which is short enough).

Reply:
Good point, I started removing dsid in favour of dataid in the code.